projects/igniteui-angular/src/lib/services/overlay/overlay.ts

```diff
@@ -143,14 +149,69 @@ export class IgxOverlayService implements OnDestroy {
         settings = Object.assign({}, this._defaultSettings, settings);
         info.settings = settings;
 
-        const eventArgs = { id, componentRef: info.componentRef, cancel: false };
+        //  show overlay in set time out. This allow us to return the id before overlay appears
+        setTimeout(() => {
```

Comment:
Unfortunately, this is a workaround. Please look for a way to do this without the `setTimeout`. I wanna see no evil and setTimeout === evil.

Reply:
we add a new method - register (we may choose and better name, e.g. attach). This method will prepare the overlay and will return the ID. Then developer can show the overlay via show(id, settings) call. We are deprecating show(component, settings) method.